common/find.h

```diff
@@ -0,0 +1,122 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_FIND_H_
+#define CARBON_COMMON_FIND_H_
+
+#include <concepts>
+#include <type_traits>
+
+#include "llvm/ADT/STLExtras.h"
+
+namespace Carbon {
+
+namespace Internal {
+
+template <typename Range>
+using RangePointerType = typename std::iterator_traits<decltype(std::begin(
+    std::declval<Range>()))>::pointer;
+
+template <typename Range>
+using RangeValueType = typename std::iterator_traits<decltype(std::begin(
+    std::declval<Range>()))>::value_type;
+
+template <typename Range, typename Pred>
+concept IsValidFindPredicate =
+    requires(const RangeValueType<Range>& elem, Pred pred) {
+      { pred(elem) } -> std::convertible_to<bool>;
+    };
+
+template <typename A, typename B>
+concept IsComparable = requires(const A& a, const B& b) {
+  { a == b } -> std::convertible_to<bool>;
+};
+
+}  // namespace Internal
+
+// Finds a value in the given `range` by comparing to `query`. Returns a
+// pointer to the value from the range on success, and nullptr if nothing is
+// found.
+//
+// This is similar to `std::find_if()` but returns a pointer to the value
+// instead of an iterator that must be tested against `end()`.
+template <typename Range, typename Query = Internal::RangeValueType<Range>>
+  requires Internal::IsComparable<Query, Internal::RangeValueType<Range>>
+constexpr auto FindOrNull(Range&& range, const Query& query)
+    -> Internal::RangePointerType<Range> {
+  auto it = llvm::find(range, query);
+  if (it != range.end()) {
+    return std::addressof(*it);
+  } else {
+    return nullptr;
+  }
+}
+
+// Finds a value in the given `range` by comparing to `query` and returns a copy
+// of it. If no match is found, returns `default_value`.
+template <typename Range, typename Query, typename Default>
+  requires Internal::IsComparable<Query, Internal::RangeValueType<Range>> &&
+           std::convertible_to<Default&&, Internal::RangeValueType<Range>> &&
+           std::copy_constructible<Internal::RangeValueType<Range>>
+constexpr auto FindOrDefault(Range&& range, const Query& query,
+                             Default default_value)
+    -> Internal::RangeValueType<Range> {
+  auto it = llvm::find(range, query);
+  if (it != range.end()) {
+    return *it;
+  } else {
+    return std::move(default_value);
+  }
+}
```

Comment:
These seem like pretty rare things to want to do -- `Contains` seems to cover the common cases. And there are no calls to either of these in this PR outside the tests. Do you think they're worth adding?

Reply:
They may be uncommon for ids, but I wonder if they won't be useful for other scenarios. Given the presence of llvm::find() I am inclined to say we keep them, rather than having to add them later. The cost to having them until used seems very low. But if you feel at all strongly, I can remove them also, let me know.

Reply:
If you have some specific examples of places where you'd want to use them, I don't mind these continuing to exist, but I would prefer not to speculatively add them if not. I think things in "common/" ought to be treated as vocabulary functions/types/whatever, so there is some (small) brain-space overhead in things existing here.

Reply:
Done